swift/extractor/SwiftOutputRewrite.cpp

```diff
@@ -0,0 +1,329 @@
+//
+// This method constructs those aliases for a module only if it comes from Xcode, which is detected
+// by the presence of an `Intermediates.noindex` directory in the module path.
+//
+// In the case of the Swift Package Manager (`swift build`) this is not needed.
+static std::vector<std::string> computeModuleAliases(llvm::StringRef modulePath,
+                                                     const std::string& targetTriple) {
+  if (modulePath.empty()) {
+    return {};
+  }
+  if (!modulePath.endswith(".swiftmodule")) {
+    return {};
+  }
+  // Deconstructing the Xcode generated path
+  //
+  // clang-format off
+  //                          intermediatesDirIndex               destinationDir (2)                              arch(5)
+  // DerivedData/FooBar/Build/Intermediates.noindex/FooBar.build/Debug-iphonesimulator/FooBar.build/Objects-normal/x86_64/FooBar.swiftmodule
+  // clang-format on
+  llvm::SmallVector<llvm::StringRef> chunks;
+  modulePath.split(chunks, '/');
+  size_t intermediatesDirIndex = 0;
+  for (size_t i = 0; i < chunks.size(); i++) {
+    if (chunks[i] == "Intermediates.noindex") {
+      intermediatesDirIndex = i;
+      break;
+    }
+  }
+  // Not built by Xcode, skipping
+  if (intermediatesDirIndex == 0) {
+    return {};
+  }
+  size_t destinationDirIndex = intermediatesDirIndex + 2;
+  size_t archIndex = intermediatesDirIndex + 5;
+  if (archIndex >= chunks.size()) {
+    return {};
+  }
+  // e.g. Debug-iphoneos, Release-iphonesimulator, etc.
+  auto destinationDir = chunks[destinationDirIndex].str();
+  auto arch = chunks[archIndex].str();
+  auto moduleNameWithExt = chunks.back();
+  auto moduleName = moduleNameWithExt.substr(0, moduleNameWithExt.find_last_of('.'));
+  std::string relocatedModulePath = chunks[0].str();
+  for (size_t i = 1; i < intermediatesDirIndex; i++) {
+    relocatedModulePath += '/' + chunks[i].str();
+  }
+  relocatedModulePath += "/Products/";
+  relocatedModulePath += destinationDir + '/';
+
+  std::vector<std::string> moduleLocations;
+
+  std::string firstCase = relocatedModulePath;
+  firstCase += moduleNameWithExt.str() + '/';
+  moduleLocations.push_back(firstCase);
+
+  std::string secondCase = relocatedModulePath;
+  secondCase += moduleName.str() + '/';
+  secondCase += moduleNameWithExt.str() + '/';
+  moduleLocations.push_back(secondCase);
+
+  std::string thirdCase = relocatedModulePath;
+  thirdCase += moduleName.str() + '/';
+  thirdCase += moduleName.str() + ".framework/Modules/";
+  thirdCase += moduleNameWithExt.str() + '/';
+  moduleLocations.push_back(thirdCase);
+
+  std::vector<std::string> aliases;
+  for (auto& location : moduleLocations) {
+    aliases.push_back(location + arch + ".swiftmodule");
+    if (!targetTriple.empty()) {
+      llvm::Triple triple(targetTriple);
+      auto moduleTriple = swift::getTargetSpecificModuleTriple(triple);
+      aliases.push_back(location + moduleTriple.normalize() + ".swiftmodule");
+    }
+  }
```

Comment:
Slightly shorter, probably clearer, involves less characters stored in separate strings as we don't store the comon `relocateModulePath` part, and seems easier to update:
```suggestion
  std::vector<std::string> moduleLocationCases = {
    "",
    moduleName.str() + '/',
    moduleName.str() + '/' + moduleName.str() + ".framework/Modules/",
  };

  std::vector<std::string> aliases;
  for (const auto& dir : moduleLocationCases) {
    auto location = relocateModulePath + '/' + dir + moduleNameWithExt.str() + '/';
    aliases.push_back(location + arch+ ".swiftmodule");
    if (!targetTriple.empty()) {
      llvm::Triple triple(targetTriple);
      auto moduleTriple = swift::getTargetSpecificModuleTriple(triple);
      aliases.push_back(location + moduleTriple.normalize() + ".swiftmodule");
    }
  }
```

Reply:
I agree it's shorter and more readable, but I'd keep it as is as I think it's too early to generalize. I.e., we may find a few other cases where this approach won't work.

Reply:
ok, but for readability I would still wrap it in a single assignment rather that assign, `+=`, `push_back` sequences
```suggestion
  std::vector<std::string> moduleLocations = {
    relocatedModulePath + moduleNameWithExt.str() + '/',
    relocatedModulePath + moduleName.str() + '/' + moduleNameWithExt.str() + '/',  
    relocatedModulePath + moduleName.str() + '/' + moduleName.str() + ".framework/Modules/" + moduleNameWithExt.str() + '/',
  }

  std::vector<std::string> aliases;
  for (auto& location : moduleLocations) {
    aliases.push_back(location + arch + ".swiftmodule");
    if (!targetTriple.empty()) {
      llvm::Triple triple(targetTriple);
      auto moduleTriple = swift::getTargetSpecificModuleTriple(triple);
      aliases.push_back(location + moduleTriple.normalize() + ".swiftmodule");
    }
  }
```